enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -254,31 +253,40 @@ impl<'a> Handler<'a> {
     pub fn read(&mut self) -> u64 {
         self.trace("read", 3);
 
-        let fd = self.aex.gpr.rdi.raw();
-        let buf = self.aex.gpr.rsi.raw() as *mut u8;
-        let size = self.aex.gpr.rdx.raw();
-
-        // Allocate some unencrypted memory.
-        let map = match self.ualloc(size) {
-            Err(errno) => return errno as u64,
-            Ok(map) => map,
+        // Allocate some unencrypted memory from the Block.
+        let mut cursor = self.block.cursor();
+        let slice: &mut [u8] = match unsafe { cursor.alloc(usize::from(self.aex.gpr.rdx)) } {
+            Ok(slice) => slice,
+            Err(_) => return -libc::EMSGSIZE as u64,
         };
 
-        unsafe {
-            // Do the syscall; replace encrypted memory with unencrypted memory.
-            let ret = self.syscall(libc::SYS_read as u64, fd, map as _, size, 0, 0, 0);
+        // Do the syscall; replace encrypted memory with unencrypted memory.
+        let req =
+            request!(libc::SYS_read => self.aex.gpr.rdi, slice, usize::from(self.aex.gpr.rdx));
```

Comment:
```suggestion
        let req = request!(libc::SYS_read => self.aex.gpr.rdi, slice, slice.len());
```

Reply:
This is defensive coding. We allocate the size of `slice` from `rdx`. But what if someone introduces a bug in the future? The size we pass to the proxy should *always* be the size of the valid memory. Otherwise we crash.

Reply:
Appreciate this!